### PR TITLE
Properly handle strings that are near a page boundary

### PR DESCRIPTION
**Pull request by shepmaster**
Thanks to the test from @borman, we now better handle this nasty edge case! @borman, if you have a second, I'd appreciate your eyes on my changes. If not, that's fine too!

I also cleaned up your submitted tests to make them a bit more Rust-like. If you have any questions, I'd be more than happy to drop some knowledge on you, in return for the gifts you gave me!

### CODE REVIEW

**Comment by shepmaster** [comment]
Of course, there's no free lunch here:

|  | Before (MB/s) | After (MB/s) | Percent |
| --- | --- | --- | --- |
| space | 6296 | 5630 | 89.4% |
| xml_delim_3 | 6118 | 5740 | 93.8% |
| xml_delim_5 | 6066 | 5022 | 82.8% |

Still orders of magnitude better than the built-in solution.

File: src/lib.rs [hunk 1]
```diff
@@ -67,29 +67,80 @@ impl AsciiChars {
...
+        let mut offset = initial_offset;
+
+        // If the string is magically aligned, skip this extra work
+        if offset != 0 {
+            unsafe {
+                let matching_bytes: usize;
+
+                asm!("pcmpestrm $$0, ($1), $2;"
+                     : // output operands
+                     "={xmm0}"(matching_bytes)
```
Comment (borman) [inline src/lib.rs]: Is it ok that you put a 128-bit SSE register in a 64-bit usize variable? I guess not.

Comment (borman) [inline src/lib.rs]: Disregard that; I forgot that *m instructions only use lower 16 bits.

File: src/lib.rs [hunk 2]
```diff
@@ -67,29 +67,80 @@ impl AsciiChars {
...
+        // starts
+        let ptr = (true_ptr as usize & !0xF) as *const u8;
 
-        loop {
+        // Find where the string really starts
+        let initial_offset = true_ptr as usize & 0xF;
+        let mut offset = initial_offset;
+
+        // If the string is magically aligned, skip this extra work
+        if offset != 0 {
```
Comment (borman) [inline src/lib.rs:98]: Wouldn't it be better to split this branch into a separate function?

**Comment by shepmaster** [comment]
> find out these fancy new ways to do things

Hmm. In the run-up to 1.0 it has been hard to keep up with all that. Personally, I skim the Rust issues and RFCs, read the subreddit, hang out in IRC, and answer a bunch of questions on Stack Overflow. [This Week in Rust](http://this-week-in-rust.org/) should be a pretty good rollup of interesting stuff though!

> & operator to cast String to &str?

FWIW, this is called "deref coercion", as you actually have to have `&String` -> `&str`. This comes from the [`Deref` trait](http://doc.rust-lang.org/std/ops/trait.Deref.html).

File: src/lib.rs (hunk 2 above)
Comment (shepmaster) [inline src/lib.rs:98]: Yeah, I had debated that myself. Two half-suggestions seems like one full suggestion to me! I've pushed a new version.

**Comment by shepmaster** [comment]
@borman if this code looks good to you, I'll give it a merge today. If I don't hear from you, I'll give it a merge today :smiling_imp:

**Comment by borman** [comment]
Looks just fine!

**Comment by shepmaster** [comment]
@borman again, I can't express how happy I am that you made these tests for us - I had no idea how to exercise that corner! 

How did you even stumble on this project anyway? I was too scared to publicize it until I got the non-SSE4.2 fallback working :-)

**Comment by borman** [comment]
Seems like I've found your github profile following a link from SO. Then, this repo caught my attention. Especially because I tried to tame these instructions myself some time ago, although that attempt was not really successful. Anyway, I ended up googling&reading specs on SSE4.2 quite a lot, and I thought that having a proper interface that just works would be really nice.

**Comment by shepmaster** [comment]
@borman that's great! I hope you have a chance to use this library for something awesome someday!
